Mathlib/RingTheory/Ideal/Over.lean

```diff
@@ -429,4 +428,129 @@ lemma map_eq_top_iff {R S} [CommRing R] [CommRing S]
 
 end IsDomain
 
+section ideal_liesOver
+
+section Semiring
+
+variable (A : Type*) [CommSemiring A] {B : Type*} [Semiring B] [Algebra A B]
+  (P : Ideal B) (p : Ideal A)
+
+/-- The ideal obtained by pulling back the ideal `P` from `B` to `A`. -/
+abbrev under : Ideal A := Ideal.comap (algebraMap A B) P
+
+theorem under_def : P.under A = Ideal.comap (algebraMap A B) P := rfl
+
+instance IsPrime.under [hP : P.IsPrime] : (P.under A).IsPrime :=
+  hP.comap (algebraMap A B)
+
+variable {A}
+
+/-- `P` lies over `p` if `p` is the preimage of `P` of the `algebraMap`. -/
+class LiesOver : Prop where
+  over : p = P.under A
+
+instance over_under : P.LiesOver (P.under A) where over := rfl
+
+theorem over_def [P.LiesOver p] : p = P.under A := LiesOver.over
+
+theorem mem_of_liesOver [P.LiesOver p] (x : A) : x ∈ p ↔ algebraMap A B x ∈ P := by
+  rw [P.over_def p]
+  rfl
+
+end Semiring
+
+section CommSemiring
+
+variable {A : Type*} [CommSemiring A] {B : Type*} [CommSemiring B] {C : Type*} [Semiring C]
+  [Algebra A B] [Algebra B C] [Algebra A C] [IsScalarTower A B C]
+  (𝔓 : Ideal C) (P : Ideal B) (p : Ideal A)
+
+@[simp]
+theorem under_under : (𝔓.under B).under A  = 𝔓.under A := by
+  simp_rw [comap_comap, ← IsScalarTower.algebraMap_eq]
+
+theorem LiesOver.trans [𝔓.LiesOver P] [P.LiesOver p] : 𝔓.LiesOver p where
+  over := by rw [P.over_def p, 𝔓.over_def P, under_under]
+
+theorem LiesOver.tower_bot [hp : 𝔓.LiesOver p] [hP : 𝔓.LiesOver P] : P.LiesOver p where
+  over := by rw [𝔓.over_def p, 𝔓.over_def P, under_under]
+
+variable (B)
+
+instance under_liesOver_of_liesOver [𝔓.LiesOver p] : (𝔓.under B).LiesOver p :=
+  LiesOver.tower_bot 𝔓 (𝔓.under B) p
+
+end CommSemiring
+
+section CommRing
+
+variable {A : Type*} [CommRing A] {B : Type*} [CommRing B] [Algebra A B]
+  (P : Ideal B) (p : Ideal A) [P.LiesOver p]
+
+/-- If `P` lies over `p`, then canonically `B ⧸ P` is a `A ⧸ p`-algebra. -/
+instance algebraQuotientOfLiesOver : Algebra (A ⧸ p) (B ⧸ P) :=
+  Ideal.Quotient.algebraQuotientOfLEComap (le_of_eq (P.over_def p))
+
+instance : IsScalarTower A (A ⧸ p) (B ⧸ P) :=
+  IsScalarTower.of_algebraMap_eq' rfl
+
+/-- `B ⧸ P` is a finite `A ⧸ p`-module if `B` is a finite `A`-module. -/
+instance _root_.Module.Finite.algebra_quotient [Module.Finite A B] :
+    Module.Finite (A ⧸ p) (B ⧸ P) :=
+  Module.Finite.of_restrictScalars_finite A (A ⧸ p) (B ⧸ P)
+
+example [Module.Finite A B] : Module.Finite (A ⧸ P.under A) (B ⧸ P) := inferInstance
+
+instance (R : Type*) {S : Type*} [CommSemiring R] [CommRing S] [Algebra R S] (I : Ideal S)
+    [h : Algebra.FiniteType R S] : Algebra.FiniteType R (S ⧸ I) :=
+  Algebra.FiniteType.trans (S := S) h inferInstance
```

Comment:
I don't find a suitable place to put it because it requires importing files related to both `Algebra.FiniteType` and the quotient of rings at the same time.

Reply:
I think it is fine to import ideal quotients in the `Algebra.FiniteType` file.